server/core/src/main/java/org/infinispan/server/core/transport/CacheInitializeInboundAdapter.java

```diff
@@ -0,0 +1,57 @@
+package org.infinispan.server.core.transport;
+
+import org.infinispan.server.core.AbstractProtocolServer;
+import org.infinispan.server.core.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+
+public class CacheInitializeInboundAdapter extends ChannelInboundHandlerAdapter {
+   private static final Log log = LogFactory.getLog(CacheInitializeInboundAdapter.class, Log.class);
+   public static final Object CACHE_INITIALIZE_EVENT = new Object();
+
+   private final AbstractProtocolServer<?> server;
+
+   public CacheInitializeInboundAdapter(AbstractProtocolServer<?> server) {
+      this.server = server;
+   }
+
+   private void initializeHandlerState(ChannelHandlerContext ctx, Runnable done) {
```

Comment:
I wonder, it might help perf if we remove this handler from the pipeline after this method has completed as it should only be needed once at initialization or active.